### PR TITLE
Gherkin Rule keyword and Protocol Buffer messages

### PR DESCRIPTION
**Pull request by aslakhellesoy**
This PR adds a `Rule` keyword to the Gherkin grammar. The `GherkinDocument` AST also has a `Rule` node. The AST itself, as well as `Source` and `Pickle` messages all use Protocol Buffers.

This PR supercedes #378 and #414, and implements #250.

TODO:
- [ ] Implement `messages` library
  - [x] Java
  - [x] JavaScript
  - [ ] .NET
  - [x] Go
  - [x] Ruby
- [ ] Update `ASTBuilder` to use `messages` and rename to `GherkinDocumentBuilder`
  - [x] Java
  - [x] JavaScript
  - [ ] .NET
  - [ ] Go
  - [x] Ruby
- [ ] Remove obsolete `event-protocol` module
  - [ ] Move docs over to `messages/README.md`

### CODE REVIEW

**Comment by aslakhellesoy** [comment]
From #378 @charlierudolph:
> Cool. @aslakhellesoy how can I help with this? I wanted to get back to the protobuf work.

It would be great if you could implement the `messages` library in Go, and then update `ast_builder.go` and `pickles.go` to use it. `events.go` and `ast.go` can be removed.

I'll crack on with the JavaScript implementation.

**Comment by aslakhellesoy** [comment]
@SabotageAndi and/or @gasparnagy are you up for doing the .NET stuff?

@brasmusson or anyone else in @cucumber/cucumber-ruby - who wants to work on the Ruby code?

**Comment by SabotageAndi** [comment]
@aslakhellesoy I am full with work for the next 2-3 month. After that, perhaps. Hopefully @gasparnagy can work on it earlier.

**Comment by bv** [comment]
I can work on Ruby
Any insights @aslakhellesoy?

File: messages/java/src/main/protobuf/gherkin.proto [hunk 1]
```diff
@@ -0,0 +1,112 @@
+syntax = "proto3";
+import "sources.proto";
+package cucumber.messages;
```
Comment (mpkorstanje) [inline messages/java/src/main/protobuf/gherkin.proto]: If this translates into a java package. Can we make it `io.cucumber.messages`?

Comment (aslakhellesoy) [reply messages/java/src/main/protobuf/gherkin.proto]: Will do!

**Comment by charlierudolph** [comment]
@aslakhellesoy trying to use docker-run locally wasn't working for me, so I had to install protoc locally for development. Is this working for you?
```
$ ../../scripts/docker-run ../../Dockerfile.cucumber-build make
go get github.com/stretchr/testify
go get github.com/golang/protobuf/protoc-gen-go
protoc --go_out=. gherkin.proto
bash: /usr/local/bin/protoc: No such file or directory
make: *** [Makefile:20: gherkin_pb.go] Error 127
```

**Comment by charlierudolph** [comment]
For updating the go gherkin:
* I think I need the subrepo for messages to exist so it can be pulled, otherwise not sure how to bring that in
* Wanted to double check if you're good removing: https://github.com/cucumber/cucumber/blob/master/gherkin/go/astjson.go

**Comment by aslakhellesoy** [comment]
Thanks a lot @charlierudolph! I have been building on my host machine (MacOS) and you're right we need `protoc` in the docker image.

Since the image is based on Alpine Linux, the protobuf version you've tried to install won't work. There is a `protobuf` APK package, but the [latest version of it](https://pkgs.alpinelinux.org/packages?name=protobuf&branch=v3.7) does not support `option ruby_package`. Alpine 3.8 has a [more recent version](https://pkgs.alpinelinux.org/packages?name=protobuf&branch=v3.8) which should support it, but Alpine 3.8 is [not dockerised yet](https://hub.docker.com/_/alpine/) :-(

I see a few options here:

* Build generated sources locally and check them in
* Build the latest protobuf on Alpine
* Switch to a different Linux

I'll experiment a little more with this.

**Pull request merged**